docs/arcadia-users-group/20231104-testing-concepts/lesson.md

```diff
@@ -0,0 +1,662 @@
+```
+* Create a new branch (`git checkout -b yourinitials/init`).
+* Use git to `add`, `commit`, and `push` the two files `mean.py` and `test_mean.py` to GitHub.
+
+### GitHub Actions
+
+[GitHub Actions](https://github.com/features/actions) is a continuous integration service provided by GitHub.
+It's integrated directly into GitHub repositories and does not require additional accounts or external services.
+Note that GitHub Actions usage is free for standard GitHub-hosted runners in public repositories, and for self-hosted runners.
+For private repositories, each GitHub account receives a certain amount of free minutes and storage for use with GitHub-hosted runners, depending on the account's plan (see [here](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration) for more information).
+
+To use GitHub Actions, create a directory called `.github` and within it, create another directory called `workflows`.
+
+```
+mkdir -p .github/workflows
+```
+
+Inside the `workflows` directory, you can create a YAML file (e.g. `ci.yml`) to define your continuous integration process:
+
+```
+name: pytest CI
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest 
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python 3.10 
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.10
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r requirements.txt
+    - name: Run tests
+      run: pytest
+```
+
+Below, we break down how this workflow works:
+* `name`: Our workflow is named `pytest CI`.
+* `on`: The workflow will be triggered to run automatically whenever commits are pushed to the repository or whenever a pull request is created or updated with new commits.
+* `jobs`: Specifies what the workflow will actually run. In this case, we specify that we want to run on the latest version of the ubuntu operating system using python version 3.10. These instructions are enough to launch a computer with python running on it. Then, we specify that we want to install dependencies from a `requirements.txt` file using pip. Lastly, we run our tests using `pytest`.
+
+
+<details>
+<summary> Running GitHub Actions workflows on multiple operating systems using a matrix</summary>
+
+Often times, developers want to check that their tests will pass not just with an ubuntu operating system and one version of Python, but with many operating systems and many versions of Python.
+This can be done using a matrix, which we demonstrate below.
+We specify lists of operating systems and python versions that we want to run our CI with, and then use a matrix call to run them all.
+
+```
+name: pytest CI
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+        python-version: [3.7, 3.8, 3.9, '3.10', '3.11'] 
```

Comment:
```suggestion
        python-version: [3.7, 3.8, 3.9, 3.10, 3.11] 
```

Reply:
these ones need to be in quotes -- it fails by testing for `3.1` otherwise. it's very weird syntax tho...

Reply:
yikes. not an excuse but I checked this w gpt4 and it got it wrong too. maybe they should all be quoted, just for consistency?

Reply:
woot tested quotes on everything and that works, so added!